### PR TITLE
CXF-8712 One-off issue with retries

### PR DESCRIPTION
**Pull request by kwin**
Immediately switch to the next alternate (address or endpoint) if the
current one has already exceeded the maxNumberOfRetries

### CODE REVIEW

File: rt/features/clustering/src/main/java/org/apache/cxf/clustering/RetryStrategy.java [hunk 1]
```diff
@@ -42,7 +42,11 @@ public List<Endpoint> getAlternateEndpoints(Exchange exchange) {
     }
 
     protected <T> T getNextAlternate(List<T> alternates) {
-        return stillTheSameAddress() ? alternates.get(0) : alternates.remove(0);
+        // is the amount of retries for the first alternate already exceeded?
+        if (!stillTheSameAddress()) {
+            alternates.remove(0);
+        }
+        return alternates.isEmpty() ? null : alternates.get(0);
```
Comment (reta) [inline rt/features/clustering/src/main/java/org/apache/cxf/clustering/RetryStrategy.java:49]: This is dangerous one, `getNextAlternate()` is not supposed to return `null` and is calling for trouble [1], we should accommodate that.

[1] https://github.com/apache/cxf/blob/master/rt/features/clustering/src/main/java/org/apache/cxf/clustering/AbstractStaticFailoverStrategy.java#L113

**Comment by reta** [comment]
Thank you for the fix, @kwin , I think you are right, there is max retries + 1 attempts actually being made, we should better respect this setting.

File: rt/features/clustering/src/main/java/org/apache/cxf/clustering/RetryStrategy.java (hunk 1 above)
Comment (kwin) [reply rt/features/clustering/src/main/java/org/apache/cxf/clustering/RetryStrategy.java:49]: Any other idea how to fix without returning null? Or do we need to change the contract for of selectNextAlternate to allow null return values?

Comment (reta) [reply rt/features/clustering/src/main/java/org/apache/cxf/clustering/RetryStrategy.java:49]: This is a great question, the suggestion I have in mind is not 100% bullet proof but I would love to hear your opinion:

```
protected <T> T getNextAlternate(List<T> alternates) {
        String lastUsedAddress = null;
        // is the amount of retries for the first alternate already exceeded?
        if (!stillTheSameAddress()) {
            lastUsedAddress  = alternates.remove(0);
        }
        return alternates.isEmpty() ? lastUsedAddress : alternates.get(0);
}
```

In this case we would only go off the rails when there are no more alternatives to try out. WDYT?

Comment (kwin) [reply rt/features/clustering/src/main/java/org/apache/cxf/clustering/RetryStrategy.java:49]: This won't work for me, as I actually use this strategy with only one endpoint, so I would always run into the edge case.

Comment (reta) [reply rt/features/clustering/src/main/java/org/apache/cxf/clustering/RetryStrategy.java:49]: > This won't work for me, as I actually use this strategy with only one endpoint, so I would always run into the edge case.

Agree. Just a suggestion, in this case you actually have an easy workaround (not a solution, just temporary shortcut): reduce the number of max retries by 1. I will try to look for a better fix meantime, will update you shortly.

Comment (kwin) [reply rt/features/clustering/src/main/java/org/apache/cxf/clustering/RetryStrategy.java:49]: Thanks, if you modify the SPI anyways maybe you can also consider my comment from https://issues.apache.org/jira/browse/CXF-8288. Would be nice to always pass the message/exchange before selecting an alternate.

Comment (reta) [reply rt/features/clustering/src/main/java/org/apache/cxf/clustering/RetryStrategy.java:49]: @kwin for this particular issue (have not looked into CXF-8288) we should be able to keep SPI unchanged by documenting `getNextAlternate` could return `null` and making sure `AbstractStaticFailoverStrategy` could deal with `null` in `selectAlternateXxx` methods.

**Comment by coheigea** [comment]
@reta What's the status of this PR?

**Comment by reta** [comment]
> @reta What's the status of this PR?

@coheigea It looks abandoned, in general we need to make changes in a few places to prevent NPEs from happening

**Comment by coheigea** [comment]
> > @reta What's the status of this PR?
> 
> @coheigea It looks abandoned, in general we need to make changes in a few places to prevent NPEs from happening

Thanks @reta , in that case I removed the fix versions from the corresponding JIRA.

**Comment by kwin** [comment]
@reta What is necessary for this PR to be merged? Just https://github.com/apache/cxf/pull/954#discussion_r895181297 or anything else?

**Comment by reta** [comment]
> @reta What is necessary for this PR to be merged? Just [#954 (comment)](https://github.com/apache/cxf/pull/954#discussion_r895181297) or anything else?

@kwin just that, thank you

File: rt/features/clustering/src/main/java/org/apache/cxf/clustering/RetryStrategy.java (hunk 1 above)
Comment (kwin) [reply rt/features/clustering/src/main/java/org/apache/cxf/clustering/RetryStrategy.java:50]: Done in https://github.com/apache/cxf/pull/954/commits/cfca0e575821f577a9034d33d9f15b1638d94be2.

**Review by reta — APPROVED** [review]

**Pull request merged**